### DISCUSSION TITLE
`certificate verify failed: unable to get local issuer certificate` - Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by fatboychummy** [Rapptz/discord.py#9403] [Q&A]
Hello!

I've been pulling my hair out at this for the last two or so days, and I am completely at a loss as to what to do.

I'm running Ubuntu 20.04 with Python versions 3.8.10 and 3.9.5 tested. I have the `certifi` library installed which points to the default `ca-certificates` store location (`/etc/ssl/certs/ca-certificates.crt`).

I have tried adding certificates both "automatically" by downloading them via crt.sh and adding them to the store via `update-ca-certificates` (though I notice they are already there on a refresh via `sudo update-ca-certificates --fresh`), and forcing them via manual edits into the cert store. So far, nothing I am doing seems to work.

I have even followed some "non-safe" advice just to test via browsing to `discord.com` and just downloading the certs from there. No dice here either.

I develop mainly on my windows machine, but this needs to be deployed to my Ubuntu server (due to other applications it needs to mesh with), and it works perfectly fine on my Windows machine.

I am at a complete loss as to what I need to do here. Every discussion and issue I have found about this is usually about Windows or Mac, and the resolution is just "Double click the certificate" or "Run Install Certificates.cmd" -- which are of course not very helpful here. Unless perhaps that Install Certificates file exists on Linux as well? I haven't been able to find it if that is the case.

Please help me, I am pulling my hair out at this and don't know where to go next. Any and all information will be useful as I continue to attempt to debug this.

<details>
  <summary>Full python traceback</summary>
  
  ```
Traceback (most recent call last):
  File "/redacted/.local/lib/python3.9/site-packages/aiohttp/connector.py", line 980, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)  # type: ignore[return-value]  # noqa
  File "/usr/lib/python3.9/asyncio/base_events.py", line 1081, in create_connection
    transport, protocol = await self._create_connection_transport(
  File "/usr/lib/python3.9/asyncio/base_events.py", line 1111, in _create_connection_transport
    await waiter
  File "/usr/lib/python3.9/asyncio/sslproto.py", line 528, in data_received
    ssldata, appdata = self._sslpipe.feed_ssldata(data)
  File "/usr/lib/python3.9/asyncio/sslproto.py", line 188, in feed_ssldata
    self._sslobj.do_handshake()
  File "/usr/lib/python3.9/ssl.py", line 944, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/redacted/redacted/bot.py", line 32, in <module>
    client.run(config.bot["token"])
  File "/redacted/.local/lib/python3.9/site-packages/discord/client.py", line 860, in run
    asyncio.run(runner())
  File "/usr/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/redacted/.local/lib/python3.9/site-packages/discord/client.py", line 849, in runner
    await self.start(token, reconnect=reconnect)
  File "/redacted/.local/lib/python3.9/site-packages/discord/client.py", line 777, in start
    await self.login(token)
  File "/redacted/.local/lib/python3.9/site-packages/discord/client.py", line 612, in login
    data = await self.http.static_login(token)
  File "/redacted/.local/lib/python3.9/site-packages/discord/http.py", line 801, in static_login
    data = await self.request(Route('GET', '/users/@me'))
  File "/redacted/.local/lib/python3.9/site-packages/discord/http.py", line 624, in request
    async with self.__session.request(method, url, **kwargs) as response:
  File "/redacted/.local/lib/python3.9/site-packages/aiohttp/client.py", line 1141, in __aenter__
    self._resp = await self._coro
  File "/redacted/.local/lib/python3.9/site-packages/aiohttp/client.py", line 536, in _request
    conn = await self._connector.connect(
  File "/redacted/.local/lib/python3.9/site-packages/aiohttp/connector.py", line 540, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/redacted/.local/lib/python3.9/site-packages/aiohttp/connector.py", line 901, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
  File "/redacted/.local/lib/python3.9/site-packages/aiohttp/connector.py", line 1206, in _create_direct_connection
    raise last_exc
  File "/redacted/.local/lib/python3.9/site-packages/aiohttp/connector.py", line 1175, in _create_direct_connection
    transp, proto = await self._wrap_create_connection(
  File "/redacted/.local/lib/python3.9/site-packages/aiohttp/connector.py", line 982, in _wrap_create_connection
    raise ClientConnectorCertificateError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorCertificateError: Cannot connect to host discord.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)')]
```
</details>

<details>
  <summary>Some sample code as a sort of "minimum reproduce-able code"</summary>
  
  ```python
# Simple echoer of bruhs for testing

import discord

import config

intents = discord.Intents.default()
intents.message_content = True

client = discord.Client(intents=intents)

@client.event
async def on_ready():
    print(f"We have logged in as {client.user}")
    channel = client.get_channel(config.bot["channel_id"]) 
    await channel.send("Bridge started.")

@client.event
async def on_message(message):
    print(f"Received message: {message}")
    if message.author == client.user:
        return
            
    if message.channel.id != config.bot["channel_id"]:
        return
            
    if message.content == "bruh" or message.content == "Bruh":
        await message.channel.send("Bruh")


if __name__ == "__main__":
    client.run(config.bot["token"]) # Errors here when trying to connect. The token is not bad -- rather the SSL is bad.

    print("Dude has stopped bro")
```
</details>

**Comment by SakuraaDevelopment** [comment]
For me, I only get this when running on Mac OS. I have no clue how to fix but just wanted to say 😆 I don't have this issue on windows

**Comment by AbstractUmbra** [comment]
Firstly, Linux works differently (regarding installation of certs) then MacOS or Windows. So that easy option doesn't exist.

Common problems with certificates you seem to have tried. I recommend forcing an install and update (`pip install -U`) of ca-certificates there too.
Also please ensure your system time is correct.

**Reply by fatboychummy** [reply]
`ca-certificates` is not a pip lib -- it is installed via apt, and is up to date. `certifi` uses it though, but that is also up to date. I've checked the system time and it looks correct as well. At least, its in a timezone which is an hour behind my current timezone and it displays as such.

... Although, I *do* notice now that the "latest" version seems to be labelled as `2021/10/16`. That feels a bit old. I'll have to look and see if there's a newer version it isn't telling me about somewhere, perhaps that has the certs I'm needing.

```
$ sudo apt upgrade ca-certificates
<... clipped ...>
ca-certificates is already the newest version (20211016ubuntu0.20.04.1).
<... clipped ...>
```

**Reply by fatboychummy** [reply]
Update on ca-certificates version: I have a friend running the same module version, and it works fine for them.

**Comment by fatboychummy** [comment]
Small context update: Python's `requests` library *is* able to connect to discord.com.

```python
import requests
requests.get("https://discord.com", verify=True)
```

This works, and returns 200. Seems `aiohttp` is doing something extremely weird. I've trimmed down the minimum viable code to the following:

```python
import asyncio
import aiohttp

async def main():
  async with aiohttp.ClientSession() as session:
    async with session.get('https://discord.com') as response:
      print('Status: ', response.status)

loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

This here is enough to get an ssl error.

I have made an issue on aiohttp's repository. Hopefully they may know more about what to do to fix this issue.

aio-libs/aiohttp#7287
